pkg/traces/traces.go

```diff
@@ -0,0 +1,99 @@
+package traces
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"runtime"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	ApplicationName           = "launcher"
+	defaultSpanName           = "launcher/unknown"
+	defaultAttributeNamespace = "unknown"
+)
+
+// StartSpan returns a new context and span, including information about the calling function
+// as appropriate. Standardizes the tracer name. The caller is always responsible for
+// ending the span. `keyVals` should be a list of pairs, where the first in the pair is a
+// string representing the attribute key and the second in the pair is the attribute value.
+func StartSpan(ctx context.Context, keyVals ...interface{}) (context.Context, trace.Span) {
+	spanName := defaultSpanName
+
+	opts := make([]trace.SpanStartOption, 0)
+
+	// Extract information about the caller to set some standard attributes (code.filepath,
+	// code.lineno, code.function) and to set more specific span and attribute names.
+	// runtime.Caller(0) would return information about `StartSpan` -- calling
+	// runtime.Caller(1) will return information about the function calling `StartSpan`.
+	pc, callerFile, callerLine, ok := runtime.Caller(1)
+	if ok {
+		opts = append(opts, trace.WithAttributes(
+			semconv.CodeFilepath(callerFile),
+			semconv.CodeLineNumber(callerLine),
+		))
+
+		// Extract the calling function name and use it to set code.function and the span name.
+		if f := runtime.FuncForPC(pc); f != nil {
+			spanName = filepath.Base(f.Name())
+			opts = append(opts, trace.WithAttributes(semconv.CodeFunction(f.Name())))
+		}
+	}
+
+	opts = append(opts, trace.WithAttributes(buildAttributes(callerFile, keyVals...)...))
+
+	return otel.Tracer(ApplicationName).Start(ctx, spanName, opts...)
+}
+
+// buildAttributes takes the given keyVals, expected to be pairs representing the key
+// and value of each attribute, and parses them appropriately, ensuring that the keys
+// have consistent and specific names. Pairs that cannot be parsed will be discarded.
+func buildAttributes(callerFile string, keyVals ...interface{}) []attribute.KeyValue {
+	callerDir := defaultAttributeNamespace
+	if callerFile != "" {
+		callerDir = filepath.Base(filepath.Dir(callerFile))
+	}
+
+	attrs := make([]attribute.KeyValue, 0)
+
+	for i := 0; i < len(keyVals); i += 2 {
+		// Keys must always be strings
+		if reflect.TypeOf(keyVals[i]).String() != "string" {
+			continue
+		}
```

Comment:
Feels weird to mix reflect and the type switch below. Either seems okay, but both feels weird. Why not
```
if _, ok := keyVals[i].(string); !ok {
     attrs = append(attrs, <bad key type at ...>)
     continue
}

Reply:
Updated!